distributed/worker_state_machine.py

```diff
@@ -2754,25 +2484,43 @@ def _handle_compute_task(self, ev: ComputeTaskEvent) -> RecsInstrs:
 
         return recommendations, instructions
 
-    def _gather_dep_done_common(self, ev: GatherDepDoneEvent) -> Iterator[TaskState]:
+    def _gather_dep_done_common(
+        self, ev: GatherDepDoneEvent
+    ) -> tuple[set[TaskState], Recs]:
         """Common code for the handlers of all subclasses of GatherDepDoneEvent.
 
-        Yields the tasks that need to transition out of flight.
+        Returns the tasks that need to transition out of flight.
+
+        See also
+        --------
+        _execute_done_common
         """
         self.comm_nbytes -= ev.total_nbytes
         keys = self.in_flight_workers.pop(ev.worker)
+        tasks = set()
+        recommendations: Recs = {}
         for key in keys:
-            ts = self.tasks[key]
-            ts.done = True
-            yield ts
+            # key *must* be still in tasks
+            ts = self.tasks.get(key)
+            assert ts, self.story(key)
+
+            ts.coming_from = None
+            self.in_flight_tasks.discard(ts)
+            if ts.state == "flight":
+                ts.done = True
+                tasks.add(ts)
+            elif ts.state == "released":
```

Comment:
How can a task be released at this point? The point of having a cancelled state in the past was to avoid this since it can lead to a couple of inconsistencies. I'm concerned that we're opening ourselves to the same problems again.

Reply:
A task that is currently being served by GatherDep or Execute can have any state, except forgotten.
The key code is in 
transition_waiting_ready (resume executing)
transition_generic_fetch (resume flight)
transition_released_forgotten (do not forget if currently in GatherDep or Execute)